### DISCUSSION TITLE
Exception handling (try catch) in gdscript

### DISCUSSION THREAD

**Discussion by Gnumaru** [godotengine/godot-proposals#4736] [Scripting]
### Describe the project you are working on

A 2d platformer

### Describe the problem or limitation you are having in your project

none

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The title is a clickbait. I DO NOT propose adding try-catch to gdscript. I would merely like to discuss alternatives.

It has already been stated that try-catch will never come into gdscript. I'm perfectly fine with that.

https://github.com/godotengine/godot/issues/3516

However, I would like to discuss with other users if there is another way of making life easier for those that need to deal with a lot of error handling. This "making life easier" could or could not involve changes to gdscript.

Currently gdscript encourages usage of int error codes for error handling. So, a user that has a series of function calls that need to be checked can make it this way:

```gdscript
# EXAMPLE 1
var error_code:int
error_code = func1()
if error_code:
	return error_code
error_code = func2()
if error_code:
	return error_code
error_code = func3()
if error_code:
	return error_code
```

The ifs can become single liners
```gdscript
# EXAMPLE 2
var error_code:int
error_code = func1()
if error_code: return error_code
error_code = func2()
if error_code: return error_code
error_code = func3()
if error_code: return error_code
```

And let's not forget that gdscript supports semicolons
```gdscript
# EXAMPLE 3
var error_code:int
error_code = func1(); if error_code: return error_code
error_code = func2(); if error_code: return error_code
error_code = func3(); if error_code: return error_code
```

If I'm not mistaken, that's the limit with current gdscript.

Try-catch will never come. That's okay. If it did come, it could be like this

```gdscript
# EXAMPLE 4
try:
	func1()
	func2()
	func3()
except err:
	return err
```

But it will never come, and that's totally fine.

So, what can be done to make the user experience more pleasant, similar to a try-catch?

Perl 5 has an interesting concept called "default variable". It is a local variable that is already declared in any context. Functions with apparently discarded return values instead have the return values assigned to the default variable. And single parameter functions called without the parameter actually use the default variable as argument.

If gdscript had a perl like default variable named _ (underscore), or any other name, like _default, we could rewrite example 3 like this:

```gdscript
# EXAMPLE 5
func1(); if: return
func2(); if: return
func3(); if: return
```

The if without an expression uses the default variable, and the return without value uses the default variable also.

But whe could go further and create a new language construct, a pseudo-function called catch(), that would be replaced by the parser with "if _default: return _default". If this pseudo-function existed, we could rewrite example 5 like this

```gdscript
# EXAMPLE 6
func1(); catch()
func2(); catch()
func3(); catch()
```

or like this

```gdscript
# EXAMPLE 7
func1()
catch()
func2()
catch()
func3()
catch()
```

or maybe catch() could receive the function call as argument, so example 6 could be rewritten like this:
```gdscript
# EXAMPLE 8
catch(func1())
catch(func2())
catch(func3())
```

since the function return values are being stored in the default variable, the value can be reused in a chain of function calls
```gdscript
# EXAMPLE 9
catch(func1())
catch(func2(_default)) # here _default contains the return value of func1()
catch(func3(_default)) # here _default contains the return value of func2()
```

But since in perl a missing function parameter uses the default variable, writing the _default as parameter for func2() and func3() would be optional.

I know that a real exception handling system would be complex, since an exception could be thrown anywhere in a code block, like in a nested function call
```gdscript
# EXAMPLE 10
func1(func2(func3(func4())))
```

But leveraging code replacement, we could implement a poor man's try-catch using the parser alone, with the limitation that it would only "catch" exceptions in "root" statements, not in nested ones.

So, a poor man's try-catch would be written like this
```gdscript
# EXAMPLE 11
try:
	func1()
	func2()
	func3()
```

and would be transpiled to the manual error handling scheme
```gdscript
# EXAMPLE 12
var error_code:int
error_code = func1()
if error_code:
	return error_code
error_code = func2()
if error_code:
	return error_code
error_code = func3()
if error_code:
	return error_code
```

Without a catch block, without finally, just a try without catch.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

already described above

### If this enhancement will not be used often, can it be worked around with a few lines of script?

already described above

### Is there a reason why this should be core and not an add-on in the asset library?

already described above

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot-proposals/issues/3194.

**Comment by me2beats** [comment]
see also
https://github.com/godotengine/godot/issues/7643
https://github.com/godotengine/godot/issues/3516

it looks like in 16 Juan said that this will not happen, but many years have passed and gdscript is now rewritten to 2.0 :))

**Comment by nonunknown** [comment]
well I think this can be already done with godot 4, since Callables are implemented you can create a `func catch(method:Callable)` and deal with that in the created function!

**Comment by dalexeev** [comment]
You can take advantage of the fact that AND and OR operations are lazy:

```gdscript
func func1():
    print(1)
    return OK

func func2():
    print(2)
    return ERR_CANT_OPEN

func func3():
    print(3)
    return ERR_ALREADY_EXISTS

func _run():
    func1() or func2() or func3()

# Prints:
# 1
# 2
```

A little longer, but in my opinion a more idiomatic version:

```gdscript
func1() == OK and func2() == OK and func3() == OK
```

There are also other options, for example, as written above, there is `Callable` (which have `bind` method). You can write your own function that will allow you to shorten the code if some fragment is often repeated.

However, I don't see the problem in Example 1: repetition is not always bad. It is bad when large pieces of code are repeated in different parts of the application. When 1-2 lines are repeated next to each other (as in your example) - this is a common thing, there is nothing to fix here. Unless you can use the following option, it is shorter:

```gdscript
var err = func1()
if not err:
    err = func2()
if not err:
    err = func3()
return err
```

Personally, I really dislike exceptions, in my opinion they cause much more problems than a few extra lines of code. Exceptions are based on the vicious idea of deferring error handling until later, somewhere up the call stack. Exceptions break the order of program execution and require tracking of potentially the entire graph of application function calls.

I prefer return codes or compound return results (with an error status).

**Comment by Kelin2025** [comment]
I might sound rude, so apologies for that, but can someone in Godot team elaborate _(or share the link if already elaborated)_ why they are against try/catch, except the "images folder" analogy _(that's basically a faulty generalization)_ and weird takes like "better crash than handle bad code"?
Tbh writing utility libraries or creating complex abstract logic becomes a huge pain without it. 
We already have Go lang, do we really want to repeat **"`if err != nil` everywhere"** story? Do we really want to rely on handmade error codes instead of having built-in entity for that? Do we really want to sacrifice debugging opportunities as well (e.g. automatically seeing custom exceptions without doing handmade loggers)? If so, why do we have signals then, they could be hand-made as well?
Also, I came across the issues and saw a lot of people that'd like to see it in GDScript, but their issues were just closed with "never gonna happen"-like answers. Godot is advertised as community-driven. So why this feature request is being strongly ignored for like 7 years?

**Reply by AThousandShips** [reply]
Please look at the linked issues, for example https://github.com/godotengine/godot/issues/7643#issuecomment-541123090

**Reply by Calinou** [reply]
See the FAQ: [Why does Godot not use exceptions?](https://docs.godotengine.org/en/stable/about/faq.html#why-does-godot-not-use-exceptions)

**Reply by Kelin2025** [reply]
Reading
> Implementing exceptions would add a significant amount of complexity

So what? I don't get it, really. Isn't languages & engines primary purpose to take complexity out from its user? I understand that it's OSS and resources are limited, but using "it's hard" as an argument looks really weird, especially since GDScript introduced A LOT of features. Like, we've got lambdas, signals, @export keyword, static typing, but exceptions - "no, we want to keep it simple"

**Reply by Kelin2025** [reply]
> ## Why does Godot not use exceptions?
> We believe games should not crash, no matter what

That's basically what try/catch is invented for. How is that an argument against it?

**Reply by AThousandShips** [reply]
Who used "it's hard" as an argument?

"Complex" =/= "hard" and "simple" =/= "easy"

**Reply by dalexeev** [reply]
I can't say anything about the position of maintainers, but in my personal opinion, exceptions are a bad thing and I avoid using them in languages where they exist. Exceptions are a bad practice, as they break the control flow like `goto` and even worse. I realize that my opinion may be biased, but the lack of exceptions is the last thing on my list of complaints about GDScript.

However, I am not suggesting that GDScript should never have exceptions. Everything is possible if the need is justified and a consensus is reached. Proposal process means that you must explain the specific problem and the proposed solution. At the same time, you should be prepared for the fact that you will be advised of an alternative solution.

Please, let's keep our emotions in check and have a constructive discussion. Emotions, personal preferences, and implied accusations do nothing to bring about the requested features.

**Reply by AThousandShips** [reply]
> How is that an argument against it?

Because of uncaught exceptions, how should they be handled except by a crash? Then they aren't exceptions?

**Reply by Kelin2025** [reply]
> Who used "it's hard" as an argument?

Replace it with "it's complex", the rest of the comment won't change. GDScript have become way more complex since the past, I don't see "it increases complexity" being a problem here

**Reply by Kelin2025** [reply]
> Proposal process means that you must explain the specific problem and the proposed solution. Everything is possible if the need is justified and a consensus is reached

We've already had a lot of issues with people discussion exceptions. People already gave explanations, but their proposals were rejected by maintainers and these issues got locked. If you take a look at those issues tho, you'll see that like/dislike ratio is far from your position. It doesn't sound like a consensus

Just by looking at the original post in this discussion you can see that people struggle without them and resort to crutchy solutions. And by "crutchy solution" I mean not just "emotions, opinions etc.", but objectively 1) trying to utilize other language features like chaining ternaries which is just wrong semantically 2) writing a lot of boilerplate code to achieve the same behavior (that's why we got signals, to reduce boilerplate and have first class support for a common pattern, right?)

And, like I said previously, the main issue here is not just "I don't like how it looks", but, for example, "editor doesn't know about it, so it can't display those handmade errors like it shows internal errors"

And given that GDScript type system is still weak and won't cover all the cases in the nearest future, we could at least throw something here:
```gdscript
@onready var damage_dealer := $DamageDealer
@onready var hit_area := $HitArea
@onready var cooldown := $Cooldown

func declare_action():
  return HitAreaPreset.new({
    damage_dealer: damage_dealer,
    hit_area: cooldown <-
  })
```
If `HitAreaPreset` takes wrong parameters. 
Because, if I'll skip `hit_area` parameter or pass something like `cooldown` instead, this means there's an error. We shouldn't try to "correct" or ignore it. It's a wrong code. It should throw and display where the error happened, like when you try to write `global_position = "foo"`. Otherwise, without enough debug information (and no built-in exceptions = not enough debug information) it might take you a day to figure out where the problem happened
You could say that "it's an XY problem, we could just add type annotations for dictionaries", but that's only one example. Even TypeScript doesn't cover all the cases, and you can't say that inventing TypeScript is "less complex" than just adding exceptions.
Not to say that there're cases when we should throw but the reason isn't resolvable on the type level

**Reply by zinnschlag** [reply]
> > How is that an argument against it?
> 
> Because of uncaught exceptions, how should they be handled except by a crash? Then they aren't exceptions?

In an application with a main loop/event structure it is trivially easy to set up exception handling in such a way that there is never an uncaught exception. With a language/runtime system that properly support exceptions and exception safe-code one can use exceptions to make applications basically crash-proof (other than things like infinite loops, multi-threading deadlocks and such).

Anyway, IMO this discussion is unlikely to get anywhere. Exception is one of the topics where there are two camps and no amount of arguing will have people move from one to the other. At least I have never managed to argue someone into changing camps.

**Reply by Kelin2025** [reply]
> Because of uncaught exceptions, how should they be handled except by a crash? Then they aren't exceptions?

Quote from the docs: "We believe games should not crash, no matter what."
So, what is it then? 
![image](https://user-images.githubusercontent.com/4208480/232888376-60bca954-20cf-4416-820d-5a73cd762474.png)
An exception, right?  
If we'd cover it with try/catch, the game won't crash. It follows the motivation, but you're against it. So it's either the motivation in FAQ is wrong or you're not consistent in your views  

I'd choose the 2nd, because "avoid crash at all costs" is a non-productive approach that creates unintuitive behaviors. And, moreover, it's not even true for Godot. Because for now we actually crash the whole thing whenever something goes wrong. And if "avoid crash at all costs" were a true motivation, it should've ignored the error instead of crashing. But that'd be a terrible approach, 'cause you won't know what went wrong. So why doing the opposite with the users code?

**Reply by Kelin2025** [reply]
> Exception is one of the topics where there are two camps and no amount of arguing will have people move from one to the other.

Sad but true. And my issue here is that the opinion of maintainers on this topic went above the opinion of community, while one of the first lines in README says it's "community-driven"

**Reply by AThousandShips** [reply]
Is it really an exception though? Not all erroneous behaviour is an exception, there's plenty of unrecoverable situations

**Reply by AThousandShips** [reply]
> went above the opinion of community

Please cite where a large portion of the community wanted this feature, you are making claims of going against the community without any backing, you are not arguing in good faith

**Reply by zinnschlag** [reply]
> Is it really an exception though? Not all erroneous behaviour is an exception, there's plenty of unrecoverable situations

This argument is kinda self-defeating. If a situation is unrecoverable wouldn't that mean that a crash/shutdown is the only option? The statement that an application shouldn't crash pretty much implies that one does not accept situations as unrecoverable.

**Reply by Kelin2025** [reply]
> You are making claims on going against the community without any backing
<details><summary>Really? What is it then?</summary>
<p>
<img src="https://user-images.githubusercontent.com/4208480/232894578-11b1468f-b50d-4401-a448-c5e35c5cc333.png" />
<img src="https://user-images.githubusercontent.com/4208480/232894625-f26c3660-64b1-41ed-893e-da73f8fc174f.png" />
<img src="https://user-images.githubusercontent.com/4208480/232894756-10a36244-d1a7-4a8b-b867-285a0affb3a9.png" />
<img src="https://user-images.githubusercontent.com/4208480/232894811-451bc39f-3996-4282-8cb4-e41fab3d8d44.png" />
<img src="https://user-images.githubusercontent.com/4208480/232894828-d7b7d756-6c73-4b7a-a615-0f4c56d6b4e8.png" />
</p>
</details>

**Reply by AThousandShips** [reply]
It doesn't say "Godot will never crash" though

> will try to recover as gracefully as possible and keep going

**Reply by AThousandShips** [reply]
Considering we have over 2000 people who have contributed to the engine alone those are not "the community", especiallly considering only 15 upvotes on the proposal, compared to plenty of other proposals that I have seen, please stop the accusations

**Reply by zinnschlag** [reply]
> will try to recover as gracefully as possible and keep going

Only only difference is the word "try", which implies that Godot tries to avoid unrecoverable situations/tries to make them recoverable (mostly successfully, in my experience). But we are running into nitpicking-territory here.

**Reply by Kelin2025** [reply]
> It doesn't say "Godot will never crash" though

It means it doesn't follow "We believe games should not crash, no matter what" belief.

It's an argument FOR having try/catch, not AGAINST it

Moreover - OK, suppose that we have people that like exceptions and that don't like. Why covering only one portion if we could cover both? Introducing a language feature doesn't mean that anyone should use it. If someone doesn't like it - np, don't use it. Without exceptions they people are fine but the others have no choice rather than "use different language"

**Reply by AThousandShips** [reply]
Because it creates complexity in maintaining it, it increases binary size, it makes the code more complicated to work

Features aren't free, they require work to implement, to maintain, to test, and they take up processing power and space, the support of the original proposal was pretty weak, as is this one considering its been open for about 10 months and have only 6 upvotes, it doesn't really look like this is a widely demanded feature

You can't really argue that its worth adding that extra stuff to the engine if it isn't widely supported

The argument "just add it and let people choose" is not an argument

**Reply by Kelin2025** [reply]
> will try to recover as gracefully as possible and keep going

The simplest and the most effective way to recover is to delegate this task to the language user. Because they know exactly how the "recovered" state should look like

**Reply by Kelin2025** [reply]
> Considering we have over 2000 people who have contributed to the engine alone those are not "the community", especiallly considering only 15 upvotes on the proposal, compared to plenty of other proposals that I have seen, please stop the accusations

Ok we can do a community poll if you want

**Reply by Kelin2025** [reply]
> Because it creates complexity in maintaining it
> Features aren't free, they require work to implement, to maintain, to test, and they take up processing power and space

So we're back to where we started. 
![image](https://user-images.githubusercontent.com/4208480/232898166-ada8c1c8-3517-423d-80be-9117b7c2a84d.png)

**Reply by AThousandShips** [reply]
Stop misrepresenting people, I never said "hard"

Be mature and argue in good faith, I am no longer interested in arguing this with you because you are unwilling to engage with others respectfully and responsibly

Good day

**Reply by Kelin2025** [reply]
I'm not misrepresenting, you're just giving inconsistent takes for the whole thread. Not only to me, worth mentioning

**Reply by Kelin2025** [reply]
And, if you don't like my personality, you can just do a twitter poll from the official godot account. Unbiased, without expressing any opinion, just casually showing some examples with different approaches (including the ones from this page) and let them vote for their favorite. If that'll be even something "30/70" for "no try catch", that'd be a big portion of a community and not just "15 dislikes"

**Reply by kireevmp** [reply]
Hello @AThousandShips
I've read through this discussion (as well as FAQ and other issues that exist on this topic). Can you please help me (and @Kelin2025) understand your position better?

Let's put the implementation complexity out of the conversation for a moment and reason about the need for exceptions in GDScript.

1. In FAQ there is a statement that `games should not crash, no matter what`, which is a valid belief. However, Godot still can crash despite the best efforts to avoid it (e.g. recover). It is also reasonable to believe that exceptions (and handling them) are a way to **avoid** crashes because exceptions can be handled. Given that, implementing exceptions would be in line with the statement from FAQ, and not implementing exceptions would be inconsistent.
2. When something goes wrong in the user-land code, Godot will try and recover from the error. However, there is a ton of '_unrecoverable situations_' where Godot fails to do that, resorting to crashing. Using the exceptions mechanism developers would provide the correct way to recover from such errors, thus improving the stability of their games. As stated before, it's developers who know **best** how to recover from each particular error, not Godot.
3. Regardless of whether exceptions are implemented or not, developers will make a similar amount of mistakes in their code per line/file/project. Making exceptions a thing essentially reduces the number of '_unrecoverable situations_' by simply catching some percentage of all errors that happen. This means that exceptions will make Godot games more stable overall, no matter what.
4. Exceptions serve as a control-flow statement that significantly reduces the amount of boilerplate code, hacks, and workarounds. Building this tool would improve the developer experience and reduce user-land code complexity, which in turn makes games easier to create and easier to debug, even if an exception happens.

Now, to answer some of the problems with exceptions that were stated.
- "Exceptions would **significantly** increase bundle size". Would you be able to share more information on that? What is 'significant' in terms of actual numbers?
- "Exceptions would complicate the GDScript engine". There is already an error system built in Godot, including stack trace. Would you say that exceptions are radically different from this implementation? It looks like they do very similar things.
- "What would happen to a game when an unhandled exception happens?". The logical answer would be - Godot can try its best and recover from this exception as it does right now. If developers did not handle it, and Godot already has a tool to try and fix it, why not give it a shot?

In short, it seems to me that exceptions would improve the overall developer and player experience, and provide consistent and **defined** behavior for games. Developers would get a tool to control the _unhappy path_ explicitly instead of relying on an unknown and not 100% reliable 'recovery' mechanism in Godot.
As a tradeoff, we might receive a drop in performance (unmeasured as of yet), an increase in bundle size (also unmeasured), and extra effort in building/supporting this functionality.

@AThousandShips @Calinou I would be very pleased to get your feedback on that.

**Reply by AThousandShips** [reply]
Sorry I'm not interested in engaging any further in this thread

**Reply by dalexeev** [reply]
In my opinion, the lack of exceptions has the advantage that you start to think about errors. Which of them are more important and which are not, how best to process them. You cannot throw exceptions on everything without thinking about where and how the error should be handled.

Exceptions are call stack based and allow you to handle an error higher up the call stack by skipping many frames. But in my opinion, this is a bad practice. You should handle the error either directly in the method where it occurred (`assert`/`breakpoint` for errors that should not be in the release version, printing to the console/log or calling a "panic" function in cases of a fatal error), or this method should return an error as a result, which should be checked by the caller.

**Reply by Kelin2025** [reply]
@dalexeev I'm sorry but your whole comment is just "I think it's bad" but verbose. You didn't give any arguments on why it's bad and why should we do like you said. And you completely ignored the whole thread where we've been giving arguments, including the ones against "returning an error as a result". 

Not having exceptions and returning errors makes us write crutches like the ones in the initial message of this discussion. Just to make it at least not that verbose. It's OK in Go that's used mostly for microservices or small utilities. But when you have a product with a lot of complex logic (like games), having each 5th line of your code  `if err != nil` is just a terrible dx. And moreover, Go at least can return two variables so it's at least type-safe and not that bad in terms of syntax. GDscript doesn't have even that. Not to say that GDScript isn't even type-safe on its own

**Reply by dalexeev** [reply]
> Go at least can return two variables so it's at least type-safe and not that bad in terms of syntax.

See these proposals:

* #6684
* #2944
* #737
* #2816
* #5130

You can already write a `Result` class extending `Resource`/`RefCounted`/`Object` with `error`/`error_code`, `error_message` and `data` properties. Or these properties and methods may be part of a utility class dedicated to this operation. And in many cases you don't even need that, since you can use `null` as a failure indicator. And also use `bool` or error codes.

I see no advantage in this code

```gdscript
func _on_file_selected(path: String) -> void:
    var importer := Importer.new()
    try:
        importer.import(path)
        set_data(importer.result)
    except e:
        show_popup(e.message)
```

over this

```gdscript
func _on_file_selected(path: String) -> void:
    var importer := Importer.new()
    importer.import(path)
    if importer.error:
        show_popup(importer.error_message)
    else:
        set_data(importer.result)
```

But exceptions allow you to defer error handling up the call stack. And I think that this is a disadvantage, not an advantage. I can explain in more detail why I think so, I have several arguments. But it takes time and I'm not sure it's worth doing in this discussion.

**Reply by Kelin2025** [reply]
The issues you mentioned are about types. And although stuff like "union types", "object types" is cool, it will cover only a portion of error cases. And I don't expect GDScript type system to become something like TypeScript, it's too complex

Also, given that maintainers have been repeatedly using _"it introduces complexity on implementing, maintaining etc"_ as an argument, just adding exceptions is no way near "complex" as implementing all of the above

Any mature language has either exceptions or a [strong type system](https://en.wikipedia.org/wiki/Strong_and_weak_typing) to minify the amount of places where it can go wrong. At least one of them. 

**Either be strict or let the developer do it for themselves.** GDScript does neither. 

Moreover, type system is not a complete replacement for the exceptions. There's stuff that isn't covered by types:
- API requests. If some server changes its response structure, type system won't save you. And it might be broken in a way that you'll know way later
- Production logging. If some user gets an unexpected crash, I'd like to see why did it happen. I'd like it to automatically send the report with a place where error happened. But we can't have tools like https://sentry.io/welcome/ , because they intercept exceptions
- Conditions that aren't related to types. For example, `param > 0` can also be a strict requirement. Type system doesn't cover it

**Reply by Kelin2025** [reply]
> I see no advantage in this code

First of all, this story isn't only about `try/catch`, but about the concept of exceptions first

Like I mentioned earlier, take a look at this snippet:
```gdscript
@onready var damage_dealer := $DamageDealer
@onready var hit_area := $HitArea
@onready var cooldown := $Cooldown

func declare_action():
  return HitAreaPreset.new({
    damage_dealer: damage_dealer,
    hit_area: cooldown <-
  })
```
_Yes, you sent a couple of issues related to this case, but I already answered it above_

We cannot type annotate objects now. But `hit_area: cooldown ` is the same as trying to do `global_position = "foo"` or `2.call_non_existing_thing()`. It's a wrong code

I could just `throw` in `HitAreaPreset` constructor and it'd automatically open a place in IDE where did that happen, with all the useful information. Instead, it will crash way later and produce an absolutely useless trace _(because it crashed 10 steps after the actual mistake is made)_

`if error:` won't help here. `if error` doesn't let IDE know if that's an error. It's only a crutch to avoid crash. We should not avoid it here

Most of the time I have to run the game and manually check if everything works fine. And if it fails, it takes a lot of time to debug. Just because we can't explicitly say to "throw here". Good luck searching through 100 perks with 150+ lines of code each


---

What about try/catch - try less trivial examples then. Introduce several conditions where it might fail. Like I said, check out Go lang and see how verbose the code with `if/else` approach is

**Reply by dalexeev** [reply]
> Like I mentioned earlier, take a look at this snippet:

I don't understand what you see as a problem in this example. Either you use static types in the constructor (`_init(damage_dealer: Combatant, hit_area: Area2D) -> void`) and GDScript checks for this, or you manually check the passed parameters in the constructor (if you don't use static types or if the type system doesn't check this, as is the case with dictionary keys).

Instead of throwing an exception, there should be an `assert`/`breakpoint` (if the error is not expected in release builds), an `error`/`warning` output to the `console`/`log` using `push_warning`/`push_error`, a call to the "panic" function (if the error is fatal) or something else. So at least you will not miss an error during development and will be able to see the call stack, variable values and other information in the debugger. In release builds, a lot of this also works (output to the console and `user://logs/`), but you won't be able to get the call stack for performance reasons.

If you need to recover from an error, then the method should return the success status of the operation, and the caller should process the returned result. I already wrote about this above.

> API requests. If some server changes its response structure, type system won't save you. And it might be broken in a way that you'll know way later

You should assume that an HTTP request may return a response with an incorrect data structure and check its validity before proceeding, rather than trying to perform actions with an unknown structure by wrapping everything in a try. To simplify structure validation, you can use validators such as JSON Schema.

> Production logging. If some user gets an unexpected crash, I'd like to see why did it happen.

This can be implemented without exception. Console output is also logged in release builds by default. You can see information about the crash of the engine in the console, but it is not logged in `user://logs/` (probably this can be fixed).

> Conditions that aren't related to types. For example, param > 0 can also be a strict requirement. Type system doesn't cover it

If the type system does not cover something, you should check it yourself. I have listed the main options above.

> if error: won't help here. if error doesn't let IDE know if that's an error. It's only a crutch to avoid crash. We should not avoid it here

I agree that you should not mask/hide the error, but I did not suggest doing so. In the examples below, you can somehow see that something went wrong.

<details>
<summary>Options</summary>

```gdscript
# Only debug builds.
func do_something(param: int) -> SomeObject:
    assert(param > 0)
    return _do_something_unsafe(param)
```

```gdscript
# Logging to console and "user://logs".
func do_something(param: int) -> SomeObject:
    if param <= 0:
        push_error("param must be > 0")
        return null
    return _do_something_unsafe(param)
```

```gdscript
# Compound return result.
func do_something(param: int) -> Result:
    if param <= 0:
        return _make_error("param must be > 0")
    return Result.new(_do_something_unsafe(param))
```

```gdscript
# Fatal error.
func do_something(param: int) -> SomeObject:
    if param <= 0:
        App.fatal("param must be > 0")
        return null
    return _do_something_unsafe(param)
```

</details>

**Reply by dalexeev** [reply]
> I could just throw in HitAreaPreset constructor and it'd automatically open a place in IDE where did that happen, with all the useful information. Instead, it will crash way later and produce an absolutely useless trace (because it crashed 10 steps after the actual mistake is made)

![](https://user-images.githubusercontent.com/47700418/233083000-54cd1540-1e24-45de-b0cd-052ca1cf54f7.png)

![](https://user-images.githubusercontent.com/47700418/233083417-77fefad1-4f5b-4786-9846-df0e60e84623.png)

**Comment by Gnumaru** [comment]
I've stated at the top that "The title is a clickbait. I DO NOT propose adding try-catch to gdscript. I would merely like to discuss alternatives."

If gdscript had type unions and a native Exception variant, we could write things such as:

```gdscript
func get_someone()->Node|Exception:
	var node:Node = get_node_or_null('../some/node')
	if node != null: return node
	return Exception.new('node not found')

func other_func()->void|Exception:
	var node:Node|Exception = get_someone()
	if node is Exception: return Exception.new('I needed some node that was not found')
	node.do_something()
```

Wich whould help a lot in implementing and using methods that may or may not suceed as expected.

Type unions could be implemented with type erasure, so that they would only validate at compile time, but the runtime would use just Variant, wich wouldn't need any change in the runtime, only in the gdscript parser/compiler.

There is already an older proposal with 98 positive reactions about type unions:

https://github.com/godotengine/godot-proposals/issues/737
